### PR TITLE
 [TEST] K-FAC for linear layers #119 

### PR DESCRIPTION
**Pull request by jabader97**

### CODE REVIEW

File: test/extensions/secondorder/hbp/test_kfac.py [hunk 1]
```diff
@@ -23,3 +32,44 @@ def test_kfac_not_supported(problem):
...
+    autograd_res = []
+    for layer, mat in zip(problem.model.parameters(), mat_list):
+        ggn_cols = []
+        for i in range(mat.size(0)):
+            e_d = mat[i, :]
+            GGN_col_i = ggn_vector_product_from_plist(loss, outputs, [layer], e_d)[0]
+            ggn_cols.append(GGN_col_i.unsqueeze(0))
+        autograd_res.append(
+            torch.cat(ggn_cols, dim=0).reshape(layer.numel(), layer.numel())
+        )
```
Comment (f-dangel) [inline test/extensions/secondorder/hbp/test_kfac.py:63]: Can you move this into `AutogradExtensions(problem).ggn_blocks()`?

File: test/extensions/secondorder/hbp/test_kfac.py [hunk 2]
```diff
@@ -23,3 +32,44 @@ def test_kfac_not_supported(problem):
...
+
+    # calculate backpack average
+    mc_samples = 200
+    backpack_average_res = []
+    backpack_kfac = BackpackExtensions(problem).kfac(mc_samples)
+    for bpr in backpack_kfac:
+        backpack_average_res.append(kfacs_to_mat(bpr))
```
Comment (f-dangel) [inline test/extensions/secondorder/hbp/test_kfac.py:70]: Shorter: `backpack_kfac_mats = [kfacs_to_mat(kfac) for kfac in backpack_kfac]`

**Review by f-dangel — CHANGES_REQUESTED** [review]

File: test/extensions/secondorder/hbp/test_kfac.py [hunk 3]
```diff
@@ -23,3 +31,48 @@ def test_kfac_not_supported(problem):
...
+    # calculate GGN
+    autograd_res = AutogradExtensions(problem).ggn_blocks()
+
+    # calculate backpack average
+    mc_samples = 6000
+    backpack_kfac = BackpackExtensions(problem).kfac(mc_samples)
+    backpack_res = [kfacs_to_mat(kfac) for kfac in backpack_kfac]
+
+    # check the values
+    check_sizes_and_values(autograd_res, backpack_res, atol=1e-1, rtol=1e-1)
```
Comment (f-dangel) [inline test/extensions/secondorder/hbp/test_kfac.py:76]: Maybe reduce to `1e-2` if it works on all problems.

File: test/extensions/secondorder/hbp/test_kfac.py [hunk 4]
```diff
@@ -23,3 +31,48 @@ def test_kfac_not_supported(problem):
...
+    # calculate GGN
+    autograd_res = AutogradExtensions(problem).ggn_blocks()
+
+    # calculate backpack average
+    mc_samples = 300000
+    backpack_kfac = BackpackExtensions(problem).kfac(mc_samples)
+    backpack_res = [kfacs_to_mat(kfac) for kfac in backpack_kfac]
+
+    # check the values
+    check_sizes_and_values(autograd_res, backpack_res, atol=1e-1, rtol=1e-1)
```
[the remark this answers is not in the document]
Comment (f-dangel) [reply test/extensions/secondorder/hbp/test_kfac.py:54]: Maybe reduce both to `1e-3`

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 3 above)
Comment (f-dangel) [reply test/extensions/secondorder/hbp/test_kfac.py:76]: TODO: Verify it works on GPU

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (f-dangel) [reply test/extensions/secondorder/hbp/test_kfac.py:54]: TODO: Verify it works on GPU

File: test/extensions/secondorder/hbp/test_kfac.py [hunk 5]
```diff
@@ -23,3 +31,48 @@ def test_kfac_not_supported(problem):
...
     problem.tear_down()
+
+
+@pytest.mark.montecarlo
+@pytest.mark.parametrize("problem", BATCH_SIZE_1_PROBLEMS, ids=BATCH_SIZE_1_IDS)
+def test_kfac_should_approx_ggn_montecarlo(problem):
+    """Check that for batch_size = 1, the K-FAC is the same as the GGN.
+
+    Args:
+        problem (ExtensionsTestProblem): Test case.
```
[the remark this answers is not in the document]
Comment (jabader97) [reply test/extensions/secondorder/hbp/test_kfac.py:44]: Fixed this in commit 
"Fixed shape of ggn blocks, type annotation, tol sizes, and commenting"

File: test/extensions/secondorder/hbp/test_kfac.py [hunk 6]
```diff
@@ -23,3 +31,48 @@ def test_kfac_not_supported(problem):
         BackpackExtensions(problem).kfac()
 
     problem.tear_down()
+
+
+@pytest.mark.montecarlo
+@pytest.mark.parametrize("problem", BATCH_SIZE_1_PROBLEMS, ids=BATCH_SIZE_1_IDS)
+def test_kfac_should_approx_ggn_montecarlo(problem):
+    """Check that for batch_size = 1, the K-FAC is the same as the GGN.
```
[the remark this answers is not in the document]
Comment (jabader97) [reply test/extensions/secondorder/hbp/test_kfac.py:39]: Fixed this in commit 
"Fixed shape of ggn blocks, type annotation, tol sizes, and commenting"

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jabader97) [reply test/extensions/secondorder/hbp/test_kfac.py]: Reduced tolerances in commit
"Fixed shape of ggn blocks, type annotation, tol sizes, and commenting"

File: test/extensions/secondorder/hbp/test_kfac.py [hunk 7]
```diff
@@ -23,3 +31,48 @@ def test_kfac_not_supported(problem):
...
+
+    problem.tear_down()
+
+
+@pytest.mark.parametrize("problem", BATCH_SIZE_1_PROBLEMS, ids=BATCH_SIZE_1_IDS)
+def test_kfac_should_approx_ggn_montecarlo_light(problem):
+    """Check that for batch_size = 1, the K-FAC is the same as the GGN.
+
+    Args:
+        problem (ExtensionsTestProblem): Test case.
```
[the remark this answers is not in the document]
Comment (jabader97) [reply test/extensions/secondorder/hbp/test_kfac.py:65]: Fixed this in commit
"Fixed shape of ggn blocks, type annotation, tol sizes, and commenting"

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 3 above)
Comment (jabader97) [reply test/extensions/secondorder/hbp/test_kfac.py]: Reduced tolerances in commit
"Fixed shape of ggn blocks, type annotation, tol sizes, and commenting"

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jabader97) [reply test/extensions/secondorder/hbp/test_kfac.py]: Regarding running on the GPU: as written currently, it did in fact not pass on the GPU. 

I looked back at the tests I ran to determine when it passed, and found an issue with a simplifying assumption I made.
Original procedure: for each testing threshold (exp: 1e-1, 1e-2, 1e-3, etc), gradually increase the number of mc_samples until the test passes, then record this as the minimum number of mc_samples required.

Incorrect assumption: if a test has passed, adding more samples will also result in the test passing.

After looking at all passed mc_samples for each tol threshold, it seems that (probably intuitively) this is not the case, and it actually enters a period of pass/fail oscillation before consistently passing as the number of samples increases.

**That being said, GPU vs CPU performance differs slightly in the exact conversion, but seems to be comparable for this problem.**

**As far as tolerance, for 3e5 samples, 1e-2 may be safer. If 1e-3 tolerance is preferred, 3e6 samples would be safer**

And I would just like to be explicit that I only tried this on the problem outlined in kfac_settings.py (seeded), so a more accurate analysis of the relationship between mc_samples and tolerance would require another round of testing (which of course I can do, if you feel it would be valuable).

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 3 above)
Comment (jabader97) [reply test/extensions/secondorder/hbp/test_kfac.py]: (See comment on non-light version of this problem for full description of the updated info)

Technically this version passed on the GPU--but barely. More comfortable passing threshold would be 1e-1 tolerance for 6e3 samples, or 6e4 samples for 1e-2 tolerance

File: test/extensions/implementation/autograd.py [hunk 8]
```diff
@@ -170,6 +180,42 @@ def diag_ggn_mc_batch(self, mc_samples: int) -> List[Tensor]:  # noqa: D102
...
+        columns = []
+
+        for i in range(param.numel()):
+            one_hot = zeros_like(param).flatten()
+            one_hot[i] = 1.0
+            one_hot = one_hot.reshape_as(param)
+
+            (column,) = ggn_vector_product_from_plist(loss, outputs, [param], one_hot)
+            columns.append(column.flatten())
+        return reshape(cat(columns), (param.numel(), param.numel()))
```
Comment (f-dangel) [inline test/extensions/implementation/autograd.py:204]: `torch.stack(columns)` should do the job.

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (f-dangel) [reply test/extensions/secondorder/hbp/test_kfac.py:44]: You can now remove `ExtensionsTestProblem` in the docstring

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (f-dangel) [reply test/extensions/secondorder/hbp/test_kfac.py:65]: You can now remove `ExtensionsTestProblem` in the docstring

**Review by f-dangel — CHANGES_REQUESTED** [review]

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (f-dangel) [reply test/extensions/secondorder/hbp/test_kfac.py]: Thanks for the detailed analysis. Indeed, increasing the samples will only improve the accuracy on average, but their can surely be individual outliers.

How about using `5e-3`, which lies between `1e-2` and `1e-3`?

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 3 above)
Comment (f-dangel) [reply test/extensions/secondorder/hbp/test_kfac.py]: Let's stick to the current tolerances then. Thanks for the report. Marking as resolved.

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jabader97) [reply test/extensions/secondorder/hbp/test_kfac.py]: Fixed in commit
"Some docstring and code cleanup"

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jabader97) [reply test/extensions/secondorder/hbp/test_kfac.py]: Changed to 5e-3 in commit
"Some docstring and code cleanup"

File: test/extensions/secondorder/hbp/test_kfac.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jabader97) [reply test/extensions/secondorder/hbp/test_kfac.py]: Fixed in commit
"Some docstring and code cleanup"

File: test/extensions/implementation/autograd.py (hunk 8 above)
Comment (jabader97) [reply test/extensions/implementation/autograd.py]: Fixed in commit
"Some docstring and code cleanup"

**Review by f-dangel — APPROVED** [review]
